src/core/ext/lb_policy/grpclb/grpclb_channel_secure.c

```diff
@@ -0,0 +1,210 @@
+    }
+  }
+
+  size_t uri_path_len;
+  char *uri_path = gpr_strjoin_sep((const char **)addr_strs, num_grpclb_addrs,
+                                   ",", &uri_path_len);
+  for (size_t i = 0; i < num_grpclb_addrs; i++) gpr_free(addr_strs[i]);
+  gpr_free(addr_strs);
+
+  char *target_uri_str = NULL;
+  gpr_asprintf(&target_uri_str, "%s:%s",
+               grpc_sockaddr_get_uri_scheme(&addresses->addresses[0].address),
+               uri_path);
+  gpr_free(uri_path);
+
+  grpc_slice_hash_table *targets_info =
+      grpc_slice_hash_table_create(num_grpclb_addrs, *targets_info_entries);
+  for (size_t i = 0; i < num_grpclb_addrs; i++) {
+    grpc_slice_unref(targets_info_entries[i]->key);
+    gpr_free(targets_info_entries[i]);
+  }
+  gpr_free(targets_info_entries);
+
+  /* Create the channel args for the channel used to talk to the LBs.
+   *
+   * We strip out the channel arg for the LB policy name, since we want
+   * to use the default (pick_first) in this case.
+   *
+   * We also strip out the channel arg for the resolved addresses, since
+   * that will be generated by the name resolver used in the LB channel.
+   * Note that the LB channel will use the sockaddr resolver, so this
+   * won't actually generate a query to DNS (or some other name service).
+   * However, the addresses returned by the sockaddr resolver will have
+   * is_balancer=false, whereas our own addresses have is_balancer=true.
+   * We need the LB channel to return addresses with is_balancer=false
+   * so that it does not wind up recursively using the grpclb LB policy,
+   * as per the special case logic in client_channel.c.
+   *
+   * We also strip out the channel arg for the server URI,
+   * since that will be different for the LB channel than for the parent
+   * channel.  (The client channel factory will re-add this arg with
+   * the right value.)
+   *
+   * Channel credentials are removed because they aren't used anywhere down the
+   * stack.
+   *
+   * Lastly, the security connector is being substituted by a newly created one
+   * based off the channel credentials with the call credentials removed. This
+   * new version is passed down to be used for secure naming checks. */
+  static const char *keys_to_remove[] = {
+      GRPC_ARG_LB_POLICY_NAME, GRPC_ARG_LB_ADDRESSES, GRPC_ARG_SERVER_URI,
+      GRPC_ARG_SECURITY_CONNECTOR, GRPC_ARG_CHANNEL_CREDENTIALS};
+
+  size_t num_args_added = 0;
+  grpc_arg args_to_add[2];
+
+  GPR_ASSERT(channel_credentials != NULL);
+  GPR_ASSERT(targets_info != NULL);
+  /* If we have channel credentials, we are in a secure channel and may want
+   * to do secure naming checks. To this end, we add the
+   * server-to-balancernames mapping and the modified channel credentials to
+   * the channel args. Both will be consumed by the security handshaker for
+   * secure naming checks */
+  args_to_add[num_args_added++] =
+      grpc_channel_targets_info_create_channel_arg(targets_info);
```

Comment:
Although we're adding this map for the specific use-case of peer authentication, which is only needed in the secure version, I am wondering whether there will be other use-cases in the future where this might be useful in the insecure version.  Would it make sense to add this channel arg in both the secure and insecure versions (in which case we can generate the hash table in grpclb.c)?

I'm not sure whether or not this makes sense.  Feel free to push back if you don't think it's worthwhile.